0000-friends_in_high_kindednesses.md

```diff
@@ -0,0 +1,293 @@
+as concrete type operators. Because higher-kinded traits cannot be defined, and
+traits cannot be implemented for type operators, it is not possible to bound
+associated type operators by traits.
+
+Even without higher-kinded traits, it could be useful to bound associated type
+operators with some sort of higher-rank syntax, as in:
+
+```rust
+T where T: StreamingIterator, for<'a> T::Item<'a>: Display
+```
+
+However, this RFC does not propose adding this feature.
+
+## Benefits of implementing only this feature before other higher-kinded polymorphisms
+
+This feature is the first 20% of higher-kinded polymorphism which is worth 50%
+of the full implementation. It is the ideal starting point, as it will enable
+many constructs while adding relatively few complicates to the type system. By
+implementing only associated type operators, we sidestep several issues:
+
+* Defining higher-kinded traits
+* Implementing traits for type operators
+* Higher order type operators
+* Type operator parameters bound by higher-kinded traits
+* Type operator parameters applied to a given type or type parameter
+
+## Proposed syntax
+
+The syntax proposed in this RFC is very similar to the syntax of associated
+types and type aliases. An advantage of this is that users less familiar with
+the intimate details of kindedness will hopefully find this feature intuitive.
+
+To declare an associated type operator, simply declare an associated type
+with parameters on the type name, as in:
+
+```rust
+trait StreamingIterator {
+    type Item<'a>;
+    ...
+}
+```
+
+Here `Item` is an associated type operator of the kind `& -> *`.
+
+To apply the associated type operator, simply use it in the position where
+a normal type operator would be used instead, as in:
+
+```rust
+trait StreamingIterator {
+    ...
+    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
+}
+```
+
+To assign the associated type operator, use the parameters in the type
+declaration on the right-hand side of the type expression, as in:
+
+```rust
+impl<T> StreamingIterator for StreamIter<T> {
+   type Item<'a> = &'a [T];
+   ...
+}
+```
+
+Note here that a slice reference has the kind `&, * -> *`, but the local type
+parameter `T` is applied to it through partial application to form a type
+operator `& -> *`. The syntax makes it clear that the unapplied parameter is
+the lifetime `'a`, because `'a` is introduced on the type Item.
+
+This has the same appearance as the declaration of type operator aliases which
+are not associated with the trait.
+
+To add a concrete bound as an associated type operator, the syntax is the same
+as adding a concrete bound of an associated type. Here, any types or lifetimes
+which are parameters to the associated type operator are omitted (not elided):
+
+```rust
+where T: StreamingIterator<Item=&[u8]>
+```
+
+`&[u8]` is not an elided form of some `&'a [u8]`, but a type operator of the
+kind `& -> *`.
```

Comment:
> I would prefer a more obvious (but more verbose) syntax: StreamingIterator<Item<'a>=&'a [u8]>. That way it’s apparent that 'a is declared as part of the trait bound, and it also nicely matches the syntax for associated item definitions.

I agree that this syntax is quite magic (and is easily the least satisfying syntax here to me), but I'm concerned about how much this looks like some sort of higher-rank syntax. If we do add an ability to make these parameters higher-rank, I am a bit concerned about the difference in meaning about `'a` between `Item<'a>=&'a [u8]` and `for<'a> Item<'a>: Display`.

>  But for bigger number of input params or non-lifetime inputs I can't image how it would work, so we need a non-elided syntax too.

You're right, the elision syntax is right out for that reason.

> By the way, I would not like to use & to refer to lifetime - it has a perfectly good meaning of reference today. 

This syntax is not internal to Rust of course. I find `'*` hard to distinguish from `*`. I think the best syntax might be the clearest, using words: `lifetime -> type`.